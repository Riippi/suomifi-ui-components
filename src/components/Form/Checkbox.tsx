import React, { Component, ReactNode } from 'react';
import classnames from 'classnames';
import {
  HtmlInput,
  HtmlLabel,
  HtmlSpan,
  HtmlDiv,
  HtmlInputProps,
} from '../../reset';

import { idGenerator } from '../../utils/uuid';
import { logger } from '../../utils/logger';

const baseClassName = 'fi-checkbox';

const checkboxBaseClassNames = {
  container: `${baseClassName}_container`,
  input: `${baseClassName}_input`,
  label: `${baseClassName}_label`,
  statusText: `${baseClassName}_status`,
  hintText: `${baseClassName}_hintText`,
};

type CheckboxVariant = 'small' | 'large';
type CheckboxStatus = 'default' | 'error' | 'disabled';

export interface CheckboxInputProps extends HtmlInputProps {
  /** State of input checkbox */
  checked?: boolean;
  /** Custom classname for the input */
  className?: string;
  'aria-label'?: string;
  'aria-labelledby'?: string;
}

export interface CheckboxProps {
  /** Controlled checked-state - user actions use onClick to change  */
  checked?: boolean;
  /** Default status of Checkbox when not using controlled 'checked' state
   * @default false
   */
  defaultChecked?: boolean;
  /** Custom classname to extend or customize */
  className?: string;
  /** Disable Checkbox. Value won't be included when submitting */
  disabled?: boolean;
  /** Event handler to execute when clicked */
  onClick?: ({ checkboxState }: { checkboxState: boolean }) => void;
  /**
   * Label element content
   */
  children?: ReactNode;
  /**
   * 'small' | 'large'
   * @default small
   */
  variant?: CheckboxVariant;
  /**
   * 'default' | 'error' | 'disabled'
   * @default default
   */
  status?: CheckboxStatus;
  /**
   * Status text to be displayed in the status text element.
   */
  statusText?: string;
  /**
   * Hint text to be displayed under the label.
   */
  hintText?: string;
  /** Pass custom props to Checkbox's input component/element */
  checkboxInputProps?: CheckboxInputProps;
  /**
   * aria-label for the HTML input-element,
   * alternatively you can define aria-labelledby with label-element id
   */
  'aria-label'?: string;
  'aria-labelledby'?: string;
  /**
   * Unique id
   * If no id is specified, one will be generated using uuid
   * @default uuidV4
   */
  id?: string;
}

export interface CheckboxState {
  checkedState: boolean;
}

export class Checkbox extends Component<CheckboxProps> {
  state = {
    checkedState: !!this.props.checked || !!this.props.defaultChecked || false,
  };

  // eslint-disable-next-line camelcase
  UNSAFE_componentWillReceiveProps(nextProps: CheckboxProps) {
    const { checked } = nextProps;
    if (!!checked) {
      this.setState({ checkedState: !!checked });
    }
  }

  handleClick = () => {
    const { checked, onClick } = this.props;
    const { checkedState } = this.state;
    if (checked === undefined) {
      this.setState({ checkedState: !checkedState });
    }
    if (!!onClick) {
      onClick({ checkboxState: !checkedState });
    }
  };

  render() {
    const {
      id: propId,
      className,
      disabled = false,
      'aria-label': ariaLabel,
      'aria-labelledby': ariaLabelledBy,
      children,
      checkboxInputProps,
      checked: dismissChecked,
      defaultChecked: dismissDefaultChecked,
      onClick: dismissOnClick,
      hintText,
      status,
      statusText,
      ...passProps
    } = this.props;
    const { checkedState } = this.state;

    const hasLabel =
      children ||
      ariaLabel ||
      ariaLabelledBy ||
      checkboxInputProps?.['aria-label'] ||
      checkboxInputProps?.['aria-labelledby'];

    if (!hasLabel) {
      logger.error(
        'Checkbox component should have a label or at least an aria-label or aria-labelledby attribute.',
      );
    }

    const id = idGenerator(propId);
    const statusTextId = `${idGenerator(propId)}-statusText`;
    const hintTextId = `${idGenerator(propId)}-hintText`;

    const newCheckboxInputProps = {
      disabled,
      id,
      'aria-label': ariaLabel,
      'aria-labelledby': ariaLabelledBy,
      checked: !!checkedState,
      className: checkboxBaseClassNames.input,
      onChange: this.handleClick,
      ...checkboxInputProps,
    };

    return (
      <HtmlDiv
        className={classnames(checkboxBaseClassNames.container, className, {})}
      >
        <HtmlInput {...newCheckboxInputProps} type="checkbox" />
        <HtmlLabel
          htmlFor={id}
          className={checkboxBaseClassNames.label}
          {...passProps}
        >
          {children}
        </HtmlLabel>
        {hintText && (
          <HtmlSpan className={checkboxBaseClassNames.hintText} id={hintTextId}>
            {hintText}
          </HtmlSpan>
        )}

        {statusText && (
          <HtmlSpan
            className={checkboxBaseClassNames.statusText}
            id={statusTextId}
          >
            {statusText}
          </HtmlSpan>
        )}
      </HtmlDiv>
    );
  }
}
