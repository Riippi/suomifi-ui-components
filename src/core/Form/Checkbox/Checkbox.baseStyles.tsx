import { css } from 'styled-components';
import {
  withSuomifiTheme,
  TokensAndTheme,
  SuomifiThemeProp,
} from '../../theme';
import { element, font } from '../../theme/reset';

const checkedStyles = ({ theme }: SuomifiThemeProp) => css`
  &.fi-checkbox--checked .fi-checkbox_label {
    &::before,
    &::after {
      border-color: ${theme.colors.brandBase};
    }
  }
`;

const errorStyles = ({ theme }: SuomifiThemeProp) => css`
  &.fi-checkbox--error {
    & .fi-checkbox_label {
      &::before,
      &::after {
        border-color: ${theme.colors.alertBase};
      }
    }
    & .fi-checkbox_status {
      color: ${theme.colors.alertBase};
    }
  }
`;

const largeVariantStyles = ({ theme }: SuomifiThemeProp) => css`
  &.fi-checkbox--large {
    & .fi-checkbox_label {
      padding-left: 40px;
      line-height: 30px;
      min-height: 30px;

      &::before {
        content: '';
        position: absolute;
        left: 0px;
        top: 0px;
        display: inline-block;
        box-sizing: border-box;
        height: 30px;
        width: 30px;
        color: ${theme.colors.depthBase};
        border: 2px solid;
        border-radius: 2px;
      }
      &::after {
        display: inline-block;
        position: absolute;
        height: 8px;
        width: 16px;
        border-left: 3px solid;
        border-bottom: 3px solid;
        border-radius: 3px;
        border-bottom-right-radius: 2px;
        border-top-left-radius: 2px;
        transform: rotate(-45deg);
        left: 6px;
        top: 6px;
      }
    }

    & .fi-checkbox_hintText {
      padding-left: 40px;
    }
  }
`;

export const baseStyles = withSuomifiTheme(
  ({ theme }: TokensAndTheme) => css`
    ${element({ theme })}
    ${font({ theme })('bodyText')}
    
    &:focus-within {
      & .fi-checkbox_label::before {
        box-shadow: 0 0 0 3px #ffbf47;
      }
    }

    & .fi-checkbox_input {
      position: absolute;
      opacity: 0;
      z-index: -9999;
      background-color: ${theme.colors.whiteBase};

      /* Will show the checkmark depending on the checked status */
      & + .fi-checkbox_label::after {
        content: none;
      }
      &:checked + .fi-checkbox_label::after {
        content: '';
      }
    }
    /*TODO: Separate size and spacings to new entity for variant use */
    & .fi-checkbox_label {
      position: relative;
      display: inline-block;
      padding-left: 26px;
      min-height: 30px;
      min-width: 30px;
      &::before {
        content: '';
        position: absolute;
        left: 0px;
        top: 5px;
        display: inline-block;
        box-sizing: border-box;
        height: 18px;
        width: 18px;
        color: ${theme.colors.depthBase};
        border: 1px solid;
        border-radius: 2px;
      }
      &::after {
        content: '';
        display: inline-block;
        position: absolute;
        height: 5px;
        width: 8px;
        border-left: 2px solid;
        border-bottom: 2px solid;
        border-radius: 2px;
        transform: rotate(-45deg);
        left: 4px;
        top: 8px;
      }
    }

    & .fi-checkbox_hintText {
      display: block;
      padding-left: 26px;
      font-size: ${theme.typography.bodyTextSmall};
      color: ${theme.colors.depthBase};
      margin-bottom: 5px;
    }

    & .fi-checkbox_status {
      display: block;
      color: ${theme.colors.blackBase};
      font-size: 14px;
      font-weight: 600;
    }
    ${largeVariantStyles({ theme })};
    ${checkedStyles({ theme })};
    ${errorStyles({ theme })};

  `,
);
