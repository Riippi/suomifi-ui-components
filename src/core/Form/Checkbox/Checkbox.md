```js
import { Checkbox } from './Checkbox';
<>
  <Checkbox
    id="1"
    defaultChecked
    hintText="This is an example hint text"
    status="checked"
  >
    This is a regular checkbox that is checked and has a hint text
  </Checkbox>

  <Checkbox
    hintText="This is an example hint text"
    status="error"
    statusText="You need to accept the terms and conditions to continue"
  >
    This is a regular checkbox with a hint text and an error message
  </Checkbox>

  <Checkbox defaultChecked disabled>
    This is a disabled checkbox
  </Checkbox>

  <Checkbox />

  <Checkbox id="2" variant="large">
    This is a large checkbox that is unchecked
  </Checkbox>

  <Checkbox
    variant="large"
    status="error"
    statusText="You need to accept the terms and conditions to continue"
    defaultChecked
  >
    This is a checked large checkbox with an error message
  </Checkbox>
</>;
```
