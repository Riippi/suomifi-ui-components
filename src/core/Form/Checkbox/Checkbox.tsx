import React, { Component } from 'react';
import { default as styled } from 'styled-components';
import { TokensProp, InternalTokensProp } from '../../theme';
import {
  Checkbox as CompCheckbox,
  CheckboxProps as CompCheckboxProps,
  CheckboxInputProps as CompCheckboxInputProps,
} from '../../../components/Form/Checkbox';
import { baseStyles } from './Checkbox.baseStyles';
import { withSuomifiDefaultProps } from '../../theme/utils';
import classnames from 'classnames';

const baseClassName = 'fi-checkbox';

const checkboxClassNames = {
  disabled: `${baseClassName}--disabled`,
  error: `${baseClassName}--error`,
  checked: `${baseClassName}--checked`,
  large: `${baseClassName}--large`,
};

export interface CheckboxProps extends CompCheckboxProps, TokensProp {}
export interface CheckboxInputProps
  extends CompCheckboxInputProps,
    TokensProp {}

const StyledCheckbox = styled(
  ({ tokens, ...passProps }: CheckboxProps & InternalTokensProp) => (
    <CompCheckbox {...passProps} />
  ),
)`
  ${(props) => baseStyles(props)}
`;

export class Checkbox extends Component<CheckboxProps> {
  state = {
    checkedState: !!this.props.checked || !!this.props.defaultChecked,
  };

  handleClick = () => {
    const { onClick } = this.props;
    const { checkedState } = this.state;
    this.setState({ checkedState: !checkedState });
    if (!!onClick) {
      onClick({ checkboxState: !checkedState });
    }
  };

  render() {
    const {
      children,
      disabled = false,
      checked,
      onClick,
      status,
      variant,
      className,
      ...passProps
    } = withSuomifiDefaultProps(this.props);
    const { checkedState } = this.state;
    return (
      <StyledCheckbox
        disabled={disabled}
        onClick={this.handleClick}
        {...passProps}
        className={classnames(baseClassName, className, {
          [checkboxClassNames.error]: status === 'error',
          [checkboxClassNames.checked]: checkedState,
          [checkboxClassNames.large]: variant === 'large',
          [checkboxClassNames.disabled]: !!disabled,
        })}
      >
        {children}
      </StyledCheckbox>
    );
  }
}
