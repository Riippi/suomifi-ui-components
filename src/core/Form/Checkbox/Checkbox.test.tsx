import React from 'react';
import { axeTest } from '../../../utils/test/axe';
import { render } from '@testing-library/react';
import { Checkbox, CheckboxProps } from './Checkbox';

const BaseCheckbox = (props: CheckboxProps) => {
  const { id, children, ...passProps } = props;
  return (
    <Checkbox id="test" {...passProps}>
      {children}
    </Checkbox>
  );
};

const RegularTestCheckbox = (
  <BaseCheckbox data-testid="regular_id">Regular</BaseCheckbox>
);

const LargeTestCheckboxWithHintText = (
  <BaseCheckbox data-testid="large_id" variant="large" hintText="Take a hint">
    Large
  </BaseCheckbox>
);

const CheckedLargeTestCheckboxWithError = (
  <BaseCheckbox
    data-testid="largeError_id"
    variant="large"
    statusText="EROR EROR"
    defaultChecked
    status="error"
  >
    Large Checked with error
  </BaseCheckbox>
);

const DisabledTestCheckbox = (
  <BaseCheckbox data-testid="reguarlDisabled_id" disabled>
    Regular disabled
  </BaseCheckbox>
);

test('Calling render with the same component on the same container does not remount', () => {
  const checkboxRendered = render(RegularTestCheckbox);
  const { getByTestId, container, rerender } = checkboxRendered;
  expect(container.firstChild).toMatchSnapshot();
  expect(getByTestId('regular_id').textContent).toBe('Regular');

  // re-render the same component with different props
  rerender(
    <BaseCheckbox data-testid="regular_id_changed">
      Regular changed
    </BaseCheckbox>,
  );
  expect(getByTestId('regular_id_changed').textContent).toBe('Regular changed');
});

test(
  'Input should not have basic accessibility issues',
  axeTest(RegularTestCheckbox),
);

test(
  'Input should not have basic accessibility issues',
  axeTest(LargeTestCheckboxWithHintText),
);

test(
  'Input should not have basic accessibility issues',
  axeTest(CheckedLargeTestCheckboxWithError),
);

test(
  'Input should not have basic accessibility issues',
  axeTest(DisabledTestCheckbox),
);
