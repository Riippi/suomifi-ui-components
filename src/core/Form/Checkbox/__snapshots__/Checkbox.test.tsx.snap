// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calling render with the same component on the same container does not remount 1`] = `
.c1 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
}

.c2::-webkit-input-placeholder {
  color: inherit;
  opacity: 0.54;
}

.c3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  max-width: 100%;
}

.c0 {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
}

.c0:focus-within .fi-checkbox_label::before {
  box-shadow: 0 0 0 3px #ffbf47;
  outline: 3px solid transparent;
}

.c0 .fi-checkbox_input {
  position: absolute;
  opacity: 0;
  z-index: -9999;
  background-color: hsl(0,0%,100%);
}

.c0 .fi-checkbox_input + .fi-checkbox_label::after {
  content: none;
}

.c0 .fi-checkbox_input:checked + .fi-checkbox_label::after {
  content: '';
}

.c0 .fi-checkbox_label {
  position: relative;
  display: inline-block;
  padding-left: 25px;
  min-height: 30px;
  min-width: 30px;
  cursor: pointer;
}

.c0 .fi-checkbox_label::before {
  content: '';
  position: absolute;
  left: 0px;
  top: 5px;
  box-sizing: border-box;
  height: 18px;
  width: 18px;
  color: hsl(202,7%,67%);
  border: 1px solid;
  border-radius: 2px;
}

.c0 .fi-checkbox_label::after {
  content: '';
  position: absolute;
  height: 5px;
  width: 8px;
  border-left: 2px solid;
  border-bottom: 2px solid;
  border-radius: 2px;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
  left: 4px;
  top: 8px;
}

.c0 .fi-checkbox_hintText {
  display: block;
  padding-left: 25px;
  font-size: font-family:'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  color: hsl(202,7%,67%);
}

.c0 .fi-checkbox_status {
  display: block;
  color: hsl(0,0%,16%);
  font-size: 14px;
  font-weight: 600;
}

.c0.fi-checkbox--large .fi-checkbox_label {
  padding-left: 40px;
  line-height: 27px;
  min-height: 30px;
}

.c0.fi-checkbox--large .fi-checkbox_label::before {
  content: '';
  position: absolute;
  left: 0px;
  top: 0px;
  box-sizing: border-box;
  height: 30px;
  width: 30px;
  color: hsl(202,7%,67%);
  border: 2px solid;
}

.c0.fi-checkbox--large .fi-checkbox_label::after {
  position: absolute;
  height: 8px;
  width: 16px;
  border-left: 4px solid;
  border-bottom: 4px solid;
  border-radius: 3px;
  border-bottom-right-radius: 2px;
  border-top-left-radius: 2px;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
  left: 6px;
  top: 6px;
}

.c0.fi-checkbox--large .fi-checkbox_hintText {
  padding-left: 40px;
}

.c0.fi-checkbox--checked .fi-checkbox_label::before,
.c0.fi-checkbox--checked .fi-checkbox_label::after {
  border-color: hsl(212,63%,45%);
}

.c0.fi-checkbox--error .fi-checkbox_label::before {
  border-color: hsl(3,59%,48%);
  border-width: 2px;
}

.c0.fi-checkbox--error .fi-checkbox_label::after {
  border-color: hsl(3,59%,48%);
}

.c0.fi-checkbox--error .fi-checkbox_hintText {
  margin-bottom: 5px;
}

.c0.fi-checkbox--error .fi-checkbox_status {
  color: hsl(3,59%,48%);
}

.c0.fi-checkbox--disabled .fi-checkbox_label {
  cursor: not-allowed;
  color: hsl(202,7%,67%);
}

.c0.fi-checkbox--disabled .fi-checkbox_label::before,
.c0.fi-checkbox--disabled .fi-checkbox_label::after {
  background-color: hsl(202,7%,97%);
  border-color: hsl(202,7%,80%);
}

<div
  class="fi-checkbox_container c0 c1"
>
  <input
    class="fi-checkbox_input c2"
    id="test"
    type="checkbox"
  />
  <label
    class="c0 fi-checkbox_label c3"
    data-testid="regular_id"
    for="test"
  >
    Regular
  </label>
</div>
`;
